components/script/dom/htmlscriptelement.rs

```diff
@@ -243,42 +366,85 @@ impl FetchResponseListener for ClassicContext {
     /// step 4-9
+    #[allow(unsafe_code)]
     fn process_response_eof(&mut self, response: Result<ResourceFetchTiming, NetworkError>) {
-        // Step 5.
-        let load = response.and(self.status.clone()).map(|_| {
-            let metadata = self.metadata.take().unwrap();
+        let (source_text, final_url) = match (response, self.status.as_ref()) {
+            (Err(err), _) => {
+                // Step 6, response is an error.
+                finish_fetching_a_classic_script(
+                    &*self.elem.root(),
+                    self.kind.clone(),
+                    self.url.clone(),
+                    Err(err),
+                );
+                return;
+            },
+            (_, Err(err)) => {
+                // Step 6, response is an error.
+                finish_fetching_a_classic_script(
+                    &*self.elem.root(),
+                    self.kind.clone(),
+                    self.url.clone(),
+                    Err(err.clone()),
+                );
+                return;
+            },
+            (Ok(_), Ok(_)) => {
+                let metadata = self.metadata.take().unwrap();
+
+                // Step 7.
+                let encoding = metadata
+                    .charset
+                    .and_then(|encoding| Encoding::for_label(encoding.as_bytes()))
+                    .unwrap_or(self.character_encoding);
+
+                // Step 8.
+                let (source_text, _, _) = encoding.decode(&self.data);
+                (source_text, metadata.final_url)
+            },
+        };
 
-            // Step 6.
-            let encoding = metadata
-                .charset
-                .and_then(|encoding| Encoding::for_label(encoding.as_bytes()))
-                .unwrap_or(self.character_encoding);
+        let elem = self.elem.root();
+        let global = elem.global();
+        let cx = global.get_cx();
+        let _ar = enter_realm(&*global);
+
+        let options =
+            unsafe { CompileOptionsWrapper::new(*cx, final_url.as_str().as_ptr() as *const i8, 1) };
+
+        let can_compile_off_thread =
+            unsafe { CanCompileOffThread(*cx, options.ptr as *const _, source_text.len()) };
+
+        if can_compile_off_thread {
+            let context = Box::into_raw(Box::new(OffThreadCompilationContext {
+                script_element: self.elem.clone(),
+                script_kind: self.kind.clone(),
+                final_url,
+                url: self.url.clone(),
+                task_source: global.networking_task_source(),
+                canceller: global.task_canceller(TaskSourceName::Networking),
+                script_text: source_text.to_string(),
+            }));
 
-            // Step 7.
-            let (source_text, _, _) = encoding.decode(&self.data);
-            ScriptOrigin::external(
+            let source_text: Vec<u16> = source_text.encode_utf16().collect();
+
+            unsafe {
+                assert!(CompileOffThread(
```

Comment:
If we use CompileOffThread1 we can pass a UTF8 string instead. We could even store the source text inside of an Arc and avoid having to clone the underlying source at all.

However, this could also be a followup issue.

Reply:
I think we can use this straightaway, without changing too much code. I'll try to see what can be done.